### ISSUE TITLE
Segfault when using vips heifsave under certain conditions

### ISSUE THREAD

**Issue by pointydev** [libvips/libvips#3509]
## Bug report

**Describe the bug**
A segfault is generated when using `vips heifsave` under certain conditions(?)

**To Reproduce**
1) Download the following images: [A1](https://eoimages.gsfc.nasa.gov/images/imagerecords/144000/144898/BlackMarble_2016_A1.jpg) [B1](https://eoimages.gsfc.nasa.gov/images/imagerecords/144000/144898/BlackMarble_2016_B1.jpg) [C1](https://eoimages.gsfc.nasa.gov/images/imagerecords/144000/144898/BlackMarble_2016_C1.jpg) [D1](https://eoimages.gsfc.nasa.gov/images/imagerecords/144000/144898/BlackMarble_2016_D1.jpg) [A2](https://eoimages.gsfc.nasa.gov/images/imagerecords/144000/144898/BlackMarble_2016_A2.jpg) [B2](https://eoimages.gsfc.nasa.gov/images/imagerecords/144000/144898/BlackMarble_2016_B2.jpg) [C2](https://eoimages.gsfc.nasa.gov/images/imagerecords/144000/144898/BlackMarble_2016_C2.jpg) [D2](https://eoimages.gsfc.nasa.gov/images/imagerecords/144000/144898/BlackMarble_2016_D2.jpg) (205.7MB total)
2) Combine using `vips arrayjoin "BlackMarble_2016_A1.jpg BlackMarble_2016_B1.jpg BlackMarble_2016_C1.jpg BlackMarble_2016_D1.jpg BlackMarble_2016_A2.jpg BlackMarble_2016_B2.jpg BlackMarble_2016_C2.jpg BlackMarble_2016_D2.jpg" test.v --across 4` (this will create a ~10.4GB intermediate file)
3) Try to save as AVIF using `vips heifsave test.v out.avif --lossless --effort 9 --Q 62`
4) Experience segfault

**Expected behavior**
The image is saved to disk as an AVIF file

**Actual behavior**
A segfault occurs in `vips_foreign_save_heif_write_block` due to:
> the value of [`VIPS_REGION_N_ELEMENTS`](https://github.com/libvips/libvips/blob/16aeae9e76999d5bd8ac2fb2c4f356bd84b6f7cb/libvips/foreign/heifsave.c#LL407C10) is larger than the total size of [this](https://github.com/libvips/libvips/blob/16aeae9e76999d5bd8ac2fb2c4f356bd84b6f7cb/libvips/foreign/heifsave.c#L403)
> it's a segfault cause it's calling `memcpy(q, p, VIPS_REGION_N_ELEMENTS(region))` and that overflows past `p` because the last param of memcpy is the number of bytes to copy

(thanks @aurieh for help debugging and explaining to me, in his actual words, "it's an out of bounds read in the memcpy")

**Screenshots**
![example](https://the.yog.zone/23ZZb2N.png)

**Environment**
 - OS: `Fedora 38 (KDE Plasma) x86_64`
 - Kernel: `6.2.15`
 - Vips: `vips-0:8.14.2-1.fc38.remi.x86_64` from [Remi's RPM repository](https://rpms.remirepo.net/fedora/38/remi/x86_64/repoview/vips.html) (since `vips-heif` is not available in official repos)

**Additional context**
This issue has been reproduced on Void Linux using vips `8.14.2` from [void-packages](https://github.com/void-linux/void-packages/blob/30924b872981b12c0a70f13f608b35dcc49a649a/srcpkgs/vips/template#L1-L3) by @aurieh.

Thanks,
Elliott

**Comment by lovell** [comment]
I've been able to reproduce this. Here's the backtrace:
```
#0  __memcpy_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:660
#1  0x00007ffff7aa80be in vips_foreign_save_heif_pack (heif=0x555555642160, q=0x7ffed5985710 <error: Cannot access memory at address 0x7ffed5985710>, p=0x7fff4d7f3720 "", ne=259200)
    at ../libvips/foreign/heifsave.c:316
#2  0x00007ffff7aa8490 in vips_foreign_save_heif_write_block (region=0x55555564b670, area=0x55555564b4a0, a=0x555555642160) at ../libvips/foreign/heifsave.c:407
#3  0x00007ffff7c4b312 in wbuffer_write (wbuffer=0x55555564b490) at ../libvips/iofuncs/sinkdisc.c:174
#4  0x00007ffff7c4b396 in wbuffer_write_thread (data=0x55555564b490, user_data=0x0) at ../libvips/iofuncs/sinkdisc.c:199
#5  0x00007ffff7c27ba4 in vips_threadset_work (pointer=0x55555564b8e0) at ../libvips/iofuncs/threadset.c:134
#6  0x00007ffff7c27651 in vips_thread_run (data=0x55555564b980) at ../libvips/iofuncs/thread.c:148
#7  0x00007ffff7940361 in g_thread_proxy (data=0x55555559ea80) at ../../../glib/gthread.c:831
#8  0x00007ffff768f18a in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:444
#9  0x00007ffff771dbd0 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
```
I think the problem relates to the calculation for output region `q` approximately here:

https://github.com/libvips/libvips/blob/16aeae9e76999d5bd8ac2fb2c4f356bd84b6f7cb/libvips/foreign/heifsave.c#L404

AVIF supports up to 65536x65536 but this example attempts to create a 86400 pixel wide image, so detecting the invalid dimensions sooner should probably fix this.

**Comment by lovell** [comment]
Possible fix for this at https://github.com/libvips/libvips/pull/3513

**Comment by kleisauke** [comment]
This should be fixed in libvips 8.14.3. Thanks for reporting this!
